src/js/CheckboxTree.js

```diff
@@ -81,129 +80,153 @@ class CheckboxTree extends React.Component {
+            if (child.value === node.value) {
+                return node;
+            }
+            return { ...child, checked: false };
+        });
 
-    onCollapseAll() {
-        this.expandAllNodes(false);
+        return { ...parent, children: newChildren };
     }
 
-    expandAllNodes(expand = true) {
-        const { onExpand } = this.props;
+    onCheck = (node) => {
+        const { nodes, onCheck } = this.props;
+        const parent = this.parents[node.value];
 
-        onExpand(
-            this.state.model.clone()
-                .expandAllNodes(expand)
-                .serializeList('expanded'),
-        );
-    }
+        if (parent.radioGroup) {
+            this.parents[node.value] = this.updateRadioSiblings(node, parent);
+        }
 
-    determineShallowCheckState(node, noCascade) {
-        const flatNode = this.state.model.getNode(node.value);
+        const root = this.updateParentNodes(node);
 
-        if (flatNode.isLeaf || noCascade) {
-            return flatNode.checked ? 1 : 0;
+        if (Array.isArray(nodes)) {
+            onCheck(node, root.children);
+        } else {
+            onCheck(node, root);
         }
+    }
 
-        if (this.isEveryChildChecked(node)) {
-            return 1;
+    onExpand = (node) => {
+        const { nodes, onExpand } = this.props;
+        const root = this.updateParentNodes(node);
+        if (Array.isArray(nodes)) {
+            onExpand(node, root.children);
+        } else {
+            onExpand(node, root);
         }
+    }
 
-        if (this.isSomeChildChecked(node)) {
-            return 2;
-        }
+    onNodeClick = (node) => {
+        const { onClick } = this.props;
+        onClick(node);
+    }
+
+    onExpandAll = () => {
+        this.expandAllNodes();
+    }
 
-        return 0;
+    onCollapseAll = () => {
+        this.expandAllNodes(false);
     }
 
-    isEveryChildChecked(node) {
-        return node.children.every(child => this.state.model.getNode(child.value).checkState === 1);
+    expandAllNodes = (expand = true) => {
+        const { nodes, onExpand } = this.props;
+        const expandNodes = (node) => {
+            if (node.children && node.children.length > 0) {
+                const children = node.children.map(expandNodes);
+                return { ...node, expanded: expand, children };
+            }
+            return node;
+        };
+
+        // walk tree and set all parent nodes expanded
+        if (Array.isArray(nodes)) {
+            const newNodes = nodes.map(expandNodes);
+            onExpand(newNodes[0], newNodes);
+        } else {
+            const root = { ...nodes };
+            const newChildren = root.children.map(expandNodes);
+            onExpand(newChildren[0], root);
+        }
     }
 
-    isSomeChildChecked(node) {
-        return node.children.some(child => this.state.model.getNode(child.value).checkState > 0);
+    isParent(node) {
+        return !!(node.children && node.children.length > 0);
```

Comment:
I think is is necessary if  `node.children` is `undefined`. For example, if it is a leaf node. In that case you would be returning `undefined` instead of `false`.